### PR TITLE
[SYCL][CUDA] Don't restore CUDA contexts

### LINKED ISSUE #4171: [SYCL] [CUDA] Invalid memory access in multithreaded application when multiple contexts used
**Describe the bug**
Invalid memory access when using multiple contexts.

**To Reproduce**

Build commands:
```
clang++ -isystem $DPCPP_HOME/build/include/sycl -isystem $CUDADIR/include -fsycl -fsycl-targets=nvptx64-nvidia-cuda-sycldevice -fsycl-unnamed-lambda reduced_2q_repro.cpp -o reduced_2q_repro.bin -Wl,-rpath,$CUDADIR/lib64: $DPCPP_HOME/build/lib/libsycl.so $CUDADIR/lib64/libcurand.so /usr/lib/x86_64-linux-gnu/libcuda.so $CUDADIR/lib64/libcudart_static.a -lpthread -ldl /usr/lib/x86_64-linux-gnu/librt.so

clang++ -DUSE_SAME_CTX -isystem $DPCPP_HOME/build/include/sycl -isystem $CUDADIR/include -fsycl -fsycl-targets=nvptx64-nvidia-cuda-sycldevice -fsycl-unnamed-lambda reduced_2q_repro.cpp -o reduced_2q_repro_same_ctx.bin -Wl,-rpath,$CUDADIR/lib64: $DPCPP_HOME/build/lib/libsycl.so $CUDADIR/lib64/libcurand.so /usr/lib/x86_64-linux-gnu/libcuda.so $CUDADIR/lib64/libcudart_static.a -lpthread -ldl /usr/lib/x86_64-linux-gnu/librt.so
```

The first command builds an application which will produce invalid memory access. The second one will produce an application which won't produce invalid memory access. The only difference between these two applications is that the former will use different contexts for two queues. The latter application will use the same context for both queues.

Source code:
```c++
#include <CL/sycl.hpp>
#include <CL/sycl/backend/cuda.hpp>
#include <cuda.h>

#ifdef USE_SAME_CTX
# define CONSTRUCT_CTX(name, dev) sycl::context name{dev}
# define CONSTRUCT_Q(name, ctx, dev, ...) sycl::queue name{ctx, dev, __VA_ARGS__}
#else
# define CONSTRUCT_CTX(name, dev)
# define CONSTRUCT_Q(name, ctx, dev, ...) sycl::queue name{dev, __VA_ARGS__}
#endif

void generate(sycl::queue &Q, sycl::buffer<float> &R) {
  Q.submit([&](sycl::handler &H) {
    auto Acc = R.get_access<sycl::access::mode::read_write>(H);

    H.host_task([=](sycl::interop_handle IH) {
      auto Ptr = reinterpret_cast<float *>(IH.get_native_mem<sycl::backend::cuda>(Acc));
      (void)Ptr;
    });
  }).wait_and_throw();
}

void test(sycl::device &D) {
  std::int64_t n = 1000;
  std::vector<float> RHost(n);
  std::vector<float> R1Host(n);

  CONSTRUCT_CTX(C, D);

  try {
    CONSTRUCT_Q(Q, C, D, [](sycl::exception_list EL) {
          for (auto EP : EL)
            std::rethrow_exception(EP);
        });
    sycl::buffer<float> R{RHost.data(), RHost.size()};

    generate(Q, R);

    Q.wait_and_throw();
  } catch (const std::exception &E) {
    std::cout << "Exception: " << E.what() << std::endl;
  }

  try {
    CONSTRUCT_Q(Q, C, D, [](sycl::exception_list EL) {
          for (auto EP : EL)
            std::rethrow_exception(EP);
        });
    sycl::buffer<float> R1{R1Host.data(), R1Host.size()};

    generate(Q, R1);

    Q.wait_and_throw();
  } catch (const std::exception &E) {
    std::cout << "Exception: " << E.what() << std::endl;
  }
}

int main() {
  sycl::device D{sycl::gpu_selector{}};

  std::cout << "Running on " << D.get_info<sycl::info::device::name>() << std::endl;

  for (size_t Idx = 0; Idx < 1; ++Idx) {
    std::cout << "Iteration " << Idx << " starts" << std::endl;
    test(D);
    std::cout << "Iteration " << Idx << " ends" << std::endl;
  }
}
```

`reduced_2q_repro.bin` will not fail. Though valgrind will report invalid memory accesses.

Payload of the application (i.e. actions before, after and inside host-task) can be modified in such a way that the application will eventually hang due to the same invalid memory access.

If the `for` loop in `main` has at least two iterations, the application is killed due to SEGFAULT which is due to the same memory access violation.

Culprit of this invalid memory access is that `cuCtxSetCurrent` frees some memory when host-task attempts to wait for own dependencies.

Valgrind's log (with `SYCL_PI_TRACE=-1`) is attached for `reduced_2q_repro.bin` with only a single iteration of `for` loop in `main`.

**Environment (please complete the following information):**

- OS: Linux
- Target device: CUDA
- DPC++ version: 
```
clang version 13.0.0 (git@github.com:intel/llvm.git 63ba1ceab4945c325485cb99b205558de6025f9a)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: ......
```
- CUDA version: 10.1

**Additional context**

[reduced_2q_repro.bin.memcheck.log](https://github.com/intel/llvm/files/6870044/reduced_2q_repro.bin.memcheck.log)

### PR DESCRIPTION
**Pull request by npmiller**
This patch fixes #4171.

The issue highlighted in that ticket is that CUDA contexts are bound to
threads and PI calls are executed both in the main thread and in threads
of the thread pool.

And to ensure the active contexts are correct the CUDA plugin uses
`ScopedContext` RAII struct to set the active context to the PI context
and restore the previous active context at the end of the PI call.

However for optimization purposes `ScopedContext` skips the context
recovery if there was no active context on the thread originally, which
means it leaves the PI context active on the thread.

In addition deleting a CUDA context only deactivates it on the thread
where it is deleted, it will stay active in other threads after being
deleted.

Which means that if you start from an application with no CUDA context
active, create a SYCL queue, run an operation then delete the SYCL
queue, the context on the current thread will be created, set active,
deactivated and deleted properly.

However it won't be deactivated in the threads of the thread pools,
which means that if we create another queue and run SYCL operations on
the thread pool again, that second queue will setup its own context in
the threads but then try to restore the deleted context from the previous
queue.

This patch aims to fix that issue by simply never restoring previous
active context, which means that PI calls from the second queue running
in the thread pool would just override the deleted context and not try
to restore it.

This should work well in SYCL only code as all the PI calls are guarded
by the `ScopedContext` and will change the active context accordingly,
in fact it may even provide performance improvement in certain
multi-context scenarios, because the current implementation would only
really prevent context switches for the first context used, this will
prevent context switches for the latest context used instead.

In CUDA interop scenarios, however it does mean that after running any
SYCL code CUDA interop code cannot make assumptions about the active
context and needs to reset it to whatever context it needs. But as far
as I'm aware, this is already the current practice in `oneMKL` and
`oneDNN`, where they also use a `ScopedContext` mechanism.

In summary this patch should:
* Fix trying to restore deleted contexts in internal SYCL threads
* May improve performance in certain multi-context scenarios
* Break assumptions on the active context for CUDA interop code

### CODE REVIEW

**Review by steffenlarsen — COMMENTED** [review]
I must have missed this discussion. It has been a while since we last touched this part of the implementation, but I think I understand the problem. I remember we have had other problems with "zombie contexts" previously.

As you correctly state in the comment, this was a performance optimization, but as you also state it is a trade-off between whether or not we prioritize the first context to be used in a given thread for the entirety of its life. The effect of this is that we instead assume that once a context has been used it is likely to be the next context to be used as well, which is arguably not a bad assumption either. It would be interesting to see the effect this has on existing code. Either way, I think the most important thing is that we don't set a context *every* time we do an operation with a context.

I expect that the potential drawbacks are outweighed by the added stability, so I think this solution is the best option for now.
That said, I am not sure I see a need for this to be a class anymore. The RAII aspect is gone and it is no longer a "scoped context". Should this be made a function instead?

**Comment by npmiller** [comment]
> I must have missed this discussion. It has been a while since we last touched this part of the implementation, but I think I understand the problem. I remember we have had other problems with "zombie contexts" previously.

There is more details on the associated ticket too, but the comment here should explain the bulk of it: #4171

> The effect of this is that we instead assume that once a context has been used it is likely to be the next context to be used as well, which is arguably not a bad assumption either.

I actually believe it should be a better assumption in all cases, because even if the first context is the most used, this would really only end up moving the context switch from the end of the work in the second context to the beginning of the work in the first context.

> It would be interesting to see the effect this has on existing code.

I tried a couple of benchmarks without seeing significant performance difference (BabelStream and Lulesh), I'm not sure if you have any other benchmark in mind.

I suppose I could try to add a couple specific multi-context tests to trip this up, and see if we can get performance results from them.

> That said, I am not sure I see a need for this to be a class anymore. The RAII aspect is gone and it is no longer a "scoped context". Should this be made a function instead?

You're right that the RAII aspect is no longer needed, I wanted to keep the patch fairly short so it's easier to see what's changed as the problem is a bit difficult to explain.

I'm happy to change it for a simple function if you prefer, although I kind of like the "Scoped" aspect even if it's not used as it conveys how this is meant to be used better, for example with a free standing function someone may be tempted to move it to just the context or queue creation.

**Comment by steffenlarsen** [comment]
> I'm not sure if you have any other benchmark in mind.

Not particularly. I'm just convinced the optimization was put in place for a reason, but maybe it is not as relevant anymore.

> I suppose I could try to add a couple specific multi-context tests to trip this up, and see if we can get performance results from them.

If it's possible with the current testing infrastructure, that would be nice. Seems like it should be possible to trigger the problem relatively consistently. Also gives you a reproducer in case you'd like to revisit the strategy in the future.

> You're right that the RAII aspect is no longer needed, I wanted to keep the patch fairly short so it's easier to see what's changed as the problem is a bit difficult to explain.
> 
> I'm happy to change it for a simple function if you prefer, although I kind of like the "Scoped" aspect even if it's not used as it conveys how this is meant to be used better, for example with a free standing function someone may be tempted to move it to just the context or queue creation.

That's fair. I do not feel strongly about it. Especially in the case that someone in the future might try their hand at other strategies.

**Comment by Ruyk** [comment]
> > I'm not sure if you have any other benchmark in mind.
> 
> Not particularly. I'm just convinced the optimization was put in place for a reason, but maybe it is not as relevant anymore.
> 

It was introduced to match the behaviour of the CUDA Runtime API, and in particular, the performance expectations from people porting code iteratively. 
This does not seem to be a major requirement anymore, so we could ignore it.

> > I suppose I could try to add a couple specific multi-context tests to trip this up, and see if we can get performance results from them.
> 
> If it's possible with the current testing infrastructure, that would be nice. Seems like it should be possible to trigger the problem relatively consistently. Also gives you a reproducer in case you'd like to revisit the strategy in the future.
> 
> > You're right that the RAII aspect is no longer needed, I wanted to keep the patch fairly short so it's easier to see what's changed as the problem is a bit difficult to explain.
> > I'm happy to change it for a simple function if you prefer, although I kind of like the "Scoped" aspect even if it's not used as it conveys how this is meant to be used better, for example with a free standing function someone may be tempted to move it to just the context or queue creation.
> 
> That's fair. I do not feel strongly about it. Especially in the case that someone in the future might try their hand at other strategies.

I feel, however, we should keep the RAII aspect on this. Even if we don't recover the context anymore, the idea of portions of functions working on a specific context remains. It also means if we have to recover the original behavior, its a single line change.

**Review by keryell — COMMENTED** [review]
Nice to see such a well described PR!
Unfortunately, almost nothing of this appears in the comments of the PR.
Is it possible to save some pieces as various comments in the code or some design information in some file header comments?

File: sycl/plugins/cuda/pi_cuda.cpp [hunk 1]
```diff
@@ -137,19 +137,15 @@ pi_result check_error(CUresult result, const char *function, int line,
...
 /// Scoped context is used across all PI CUDA plugin implementation
-/// to activate the PI Context on the current thread, matching the
-/// CUDA driver semantics where the context used for the CUDA Driver
-/// API is the one active on the thread.
-/// The implementation tries to avoid replacing the CUcontext if it cans
+/// to activate the PI Context on the current thread.
+/// The implementation tries to avoid replacing the CUcontext if it cans.
 class ScopedContext {
   pi_context placedContext_;
   CUcontext original_;
```
[the remark this answers is not in the document]
Comment (npmiller) [reply sycl/plugins/cuda/pi_cuda.cpp:145]: They could be moved into the constructor, but if we keep the RAII class anyway I think it's fine to leave them here

[the remark this answers is not in the document]
Comment (steffenlarsen) [reply sycl/plugins/cuda/pi_cuda.cpp:145]: If they don't serve a purpose anymore I would prefer that we remove them. Keeping it as a RAII class I can see a future benefit in, i.e. to simplify the process of reimplementing the RAII behavior without having to change the usage too much. However, if any future implementation needs these members they are not difficult nor disruptive to reintroduce.

[the remark this answers is not in the document]
Comment (npmiller) [reply sycl/plugins/cuda/pi_cuda.cpp]: I have updated it to cleanup the class

**Comment by npmiller** [comment]
> Nice to see such a well described PR!
> Unfortunately, almost nothing of this appears in the comments of the PR.
> Is it possible to save some pieces as various comments in the code or some design information in some file header comments?

Good point, I have updated it to flesh out the comment on the class a bit more, I think it should be hitting all the important points without being quite as verbose as this PR's description, what do you think?

**Review by steffenlarsen — DISMISSED** [review]
LGTM. @npmiller Are you planning on adding the aforementioned test?

**Comment by npmiller** [comment]
> LGTM. @npmiller Are you planning on adding the aforementioned test?

I had a look but this is actually trickier to test than I expected as even the original snippet doesn't fail deterministically, also I wonder if a test like that should go in `llvm-test-suite` now that the `on-device` tests moved there. So it might be best to merge this as-is and maybe just add more generic multi-context testing later on.

**Comment by bader** [comment]
Is it possible to reproduce this bug using [CUDA plug-in unit tests](https://github.com/intel/llvm/tree/sycl/sycl/unittests/pi/cuda)?

**Comment by npmiller** [comment]
> Is it possible to reproduce this bug using [CUDA plug-in unit tests](https://github.com/intel/llvm/tree/sycl/sycl/unittests/pi/cuda)?

Oh right, That might be a good place for it! I'll have another look at testing this, it might be easier calling the PI functions directly.

**Comment by keryell** [comment]
> Good point, I have updated it to flesh out the comment on the class a bit more, I think it should be hitting all the important points without being quite as verbose as this PR's description, what do you think?

Thanks, that looks good.

**Comment by npmiller** [comment]
Alright, I've just pushed 3 tests, two checking the active contexts in regular scenarios and one re-creating the thread pool scenario but in a simpler one thread test.

File: sycl/unittests/pi/cuda/test_contexts.cpp [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+  void TearDown() override {}
+
+  CudaContextsTest() = default;
+
+  ~CudaContextsTest() = default;
+};
+
+TEST_F(CudaContextsTest, ContextLifetime) {
+  // start with no active context
+  cuCtxSetCurrent(nullptr);
```
Comment (steffenlarsen) [inline sycl/unittests/pi/cuda/test_contexts.cpp:66]: A bit nitpicky, but from the [documentation](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__CTX.html#group__CUDA__CTX_1gbe562ee6258b4fcc272ca6478ca2a2f7):

> If there exists a CUDA context stack on the calling CPU thread, this will replace the top of that stack with ctx. If ctx is NULL then this will be equivalent to popping the top of the calling CPU thread's CUDA context stack (or a no-op if the calling CPU thread's CUDA context stack is empty).

Generally that should not matter, but if we want to be safe this should make this call until [`cuCtxGetCurrent`](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__CTX.html#group__CUDA__CTX_1g8f13165846b73750693640fb3e8380d0) returns NULL.

**Review by steffenlarsen — APPROVED** [review]
New tests LGTM. Minor comment.

File: sycl/unittests/pi/cuda/test_contexts.cpp (hunk 2 above)
Comment (npmiller) [reply sycl/unittests/pi/cuda/test_contexts.cpp:66]: Oh, that's a good point I didn't think about that, I got it from `test_commands.cpp` and it's also done like that in `test_mem_obj.cpp`

Comment (npmiller) [reply sycl/unittests/pi/cuda/test_contexts.cpp:66]: I'm tempted to leave it like that for this PR and do a small follow up PR with a `clearCuCtxt` utility function and fixing the other tests as well if that works for you

Comment (steffenlarsen) [reply sycl/unittests/pi/cuda/test_contexts.cpp:66]: I'm okay with that. 👍

**Review by alexbatashev — APPROVED** [review]
